File: Avery Lamp/SkillsScene.m
//
//  SkillsScene.m
//  Avery Lamp
//
//  Created by Avery Lamp on 4/7/15.
//  Copyright (c) 2015 Avery Lamp. All rights reserved.
//

#import "SkillsScene.h"
#import <CoreMotion/CoreMotion.h>

@interface SkillsScene()
@property CGSize screenSize;
@property CMMotionManager*motionManager;
@end

static const uint32_t icon = 0x1 <<0;

@implementation SkillsScene

-(void)didMoveToView:(SKView *)view{
    CGRect screenBounds = [[UIScreen mainScreen] bounds];
    CGFloat screenScale = [[UIScreen mainScreen] scale];
    CGSize screenSize = CGSizeMake(screenBounds.size.width * screenScale, screenBounds.size.height * screenScale);
    self.screenSize = screenSize;
    
    SKLabelNode *label = [[SKLabelNode alloc]initWithFontNamed:@"Hoefler Text"];
    label.text = @"Skills";
    label.name = @"Welcome Label";
    label.position = CGPointMake(screenSize.width/2,screenSize.height * 5 /6);
    label.horizontalAlignmentMode = SKLabelHorizontalAlignmentModeCenter;
    label.fontSize = 100.0f;
    label.fontColor = [UIColor blackColor];
    label.zPosition = -1;
    [self addChild:label];
    
    
    
    SKPhysicsBody* borderBody = [SKPhysicsBody bodyWithEdgeLoopFromRect:CGRectMake(0, 0, screenSize.width, screenSize.height)];
    self.physicsBody = borderBody;
    
    SKNode *ObjC = [self nodeWithName:@"Objective-C" andRadius:100 andColor:[UIColor colorWithRed:0.800f green:0.820f blue:0.671f alpha:1.00f] andTextColor:[UIColor blackColor] fontSize:36];

    SKNode *java = [self nodeWithName:@"Java" andRadius:100 andColor:[UIColor colorWithRed:0.325f green:0.369f blue:0.231f alpha:1.00f]andTextColor:[UIColor whiteColor]fontSize:44];

    SKNode *videoPro = [self nodeWithName:@"Video" andRadius:80 andColor:[UIColor colorWithRed:0.518f green:0.173f blue:0.090f alpha:1.00f]andTextColor:[UIColor whiteColor]fontSize:40];

    SKNode *photography = [self nodeWithName:@"Photography" andRadius:75 andColor:[UIColor colorWithRed:0.235f green:0.043f blue:0.137f alpha:1.00f] andTextColor:[UIColor whiteColor]fontSize:26];

    SKNode *squash = [self nodeWithName:@"Squash" andRadius:130 andColor:[UIColor colorWithRed:0.902f green:0.557f blue:0.220f alpha:1.00f]andTextColor:[UIColor blackColor] fontSize:50];

    SKNode *bassoon = [self nodeWithName:@"Bassoon" andRadius:70 andColor:[UIColor colorWithRed:0.239f green:0.710f blue:0.835f alpha:1.00f]andTextColor:[UIColor blackColor] fontSize:34];

    SKNode *cad = [self nodeWithName:@"CAD" andRadius:70 andColor:[UIColor colorWithRed:0.788f green:0.722f blue:0.631f alpha:1.00f]andTextColor:[UIColor blackColor] fontSize:34];

    SKNode *robotics = [self nodeWithName:@"Robotics" andRadius:90 andColor:[UIColor colorWithRed:0.451f green:0.663f blue:0.392f alpha:1.00f]andTextColor:[UIColor blackColor] fontSize:38];
    
    [self addChild:robotics];
    [self addChild:java];
    [self addChild:cad];
    [self addChild:videoPro];
    [self addChild:photography];
    [self addChild:bassoon];
    [self addChild:squash];
    [self addChild:ObjC];
    
    UIButton *quitButton = [[UIButton alloc]initWithFrame:CGRectMake(0, 0, 80,80 )];
    quitButton.backgroundColor = [UIColor redColor];
    [quitButton addTarget:self action:@selector(exitVC) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:quitButton];
    
}
-(void)exitVC{
    [self.VC dismissViewControllerAnimated:YES completion:^{

    }];
}

-(SKNode *)nodeWithName:(NSString *)name andRadius:(int)radius andColor:(UIColor*)color andTextColor:(UIColor *)textColor fontSize:(int)fontSizze{
    SKNode *node = [[SKNode alloc]init];
    node.name = name;
    node.position = CGPointMake(self.screenSize.width * 2/4, self.screenSize.height* 1 /2);
    node.physicsBody = [SKPhysicsBody bodyWithCircleOfRadius:radius];
    node.physicsBody.categoryBitMask = icon;
    node.physicsBody.contactTestBitMask = icon;
    node.physicsBody.affectedByGravity = YES;
    node.physicsBody.allowsRotation = YES;
    

    SKShapeNode *circle = [[SKShapeNode alloc] init];
    CGRect rect=  CGRectMake(-radius, -radius, 2*radius, 2*radius);
    
    circle.name = @"Green Node";
    circle.path = [UIBezierPath bezierPathWithOvalInRect:rect].CGPath;
    circle.fillColor = [SKColor colorWithCGColor:color.CGColor];
    circle.strokeColor = nil;
    circle.position = CGPointMake(0, 0);
    [node addChild:circle];
    
    SKLabelNode *label = [[SKLabelNode alloc]initWithFontNamed:@"ArialMT"];
    label.text = name;
    label.fontSize = fontSizze;
    label.fontColor = textColor;
    label.horizontalAlignmentMode = SKLabelHorizontalAlignmentModeCenter;
    label.verticalAlignmentMode = SKLabelVerticalAlignmentModeCenter;
    [node addChild:label];
    return node;

}

-(void)update:(NSTimeInterval)currentTime{
    
    if (_motionManager== nil) {
        _motionManager=[[CMMotionManager alloc]init];
        _motionManager.deviceMotionUpdateInterval=0.05;
        [_motionManager startDeviceMotionUpdates];
    }
    
    CMAcceleration gravity=  _motionManager.deviceMotion.gravity;
    self.physicsWorld.gravity = CGVectorMake(gravity.x*20, gravity.y*20);
    
    
}


@end

File: Avery Lamp/SkillsScene.swift
//
//  SkillsScene.swift
//  Avery Lamp
//
//  Created by Avery Lamp on 4/25/15.
//  Copyright (c) 2015 Avery Lamp. All rights reserved.
//

import UIKit
import SpriteKit

class SkillsScene: SKScene {
    
    @objc var VC : UIViewController? = nil
    
    var allNodes = [SKShapeNode]()
    
    override func didMoveToView(view: SKView) {
        super.didMoveToView(view)
        let scale = UIScreen.mainScreen().scale
        let screenSize: CGRect = UIScreen.mainScreen().bounds
        let screenWidth = screenSize.width * scale
        let screenHeight = screenSize.height * scale
        
        var exitButton = UIButton(frame: CGRectMake((screenWidth)/2 - 60, 20, 40, 40))
        //        exitButton.backgroundColor = UIColor.redColor()
        exitButton.setBackgroundImage(UIImage(named: "exitButton"), forState: UIControlState.Normal)
        exitButton.addTarget(self, action: "ExitVC", forControlEvents:UIControlEvents.TouchUpInside)
        self.view?.addSubview(exitButton)
        
        
        var skillsLabel = SKLabelNode(text: "Skills")
        skillsLabel.fontName = "DINAlternate-Bold"
        skillsLabel.fontSize = 130
        skillsLabel.fontColor = UIColor.blackColor()
        skillsLabel.horizontalAlignmentMode = SKLabelHorizontalAlignmentMode.Center
        skillsLabel.verticalAlignmentMode = SKLabelVerticalAlignmentMode.Center
        skillsLabel.position = CGPointMake(screenWidth/2, screenHeight - 125)
        self.addChild(skillsLabel)
        
        var emitter = SKEmitterNode(fileNamed: "BackgroundParticle.sks")
        emitter.position = skillsLabel.position
        self.addChild(emitter)
        
        

        
        let bottomLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 1 / 5)
        let bottomCenter = CGPointMake(screenWidth * 2/4, screenHeight * 1 / 5)
        let bottomRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 1 / 5)
        let middleLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 2 / 5)
        let middleCenter = CGPointMake(screenWidth * 2/4, screenHeight * 2 / 5)
        let middleRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 2 / 5)
        let topLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 3 / 5)
        let topCenter = CGPointMake(screenWidth * 2/4, screenHeight * 3 / 5)
        let topRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 3 / 5)
        
        allPositions.append(topLeft)
        allPositions.append(topCenter)
        allPositions.append(topRight)
        allPositions.append(middleLeft)
        allPositions.append(middleCenter)
        allPositions.append(middleRight)
        allPositions.append(bottomLeft)
        allPositions.append(bottomCenter)
        allPositions.append(bottomRight)
        
        self.physicsBody = SKPhysicsBody(edgeLoopFromRect: CGRectMake(0, 0, screenWidth, screenHeight))
        
        
        
        
        let radius = screenWidth / 8
        var test =  nodeWithSkill("test", color: UIColor.greenColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        test.position = topLeft
        self.addChild(test)
        
        var java =  nodeWithSkill("Java", color: UIColor.redColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        java.position = topCenter
        self.addChild(java)
        
        var objectiveC =  nodeWithSkill("Objective-C", color: UIColor.greenColor(), fontSize: 40, fontColor: UIColor.blueColor(),radius: radius )
        objectiveC.position = topRight
        self.addChild(objectiveC)
        
        var video =  nodeWithSkill("Video", color: UIColor.blackColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        video.position = middleLeft
        self.addChild(video)
        
        var photoshop =  nodeWithSkill("Photoshop", color: UIColor.purpleColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        photoshop.position = middleCenter
        self.addChild(photoshop)
        
        var cad =  nodeWithSkill("CAD", color: UIColor.orangeColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        cad.position = middleRight
        self.addChild(cad)
        
        var cad2 =  nodeWithSkill("CAD2", color: UIColor.darkGrayColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        cad2.position = bottomLeft
        self.addChild(cad2)
        
        var cad3 =  nodeWithSkill("CAD3", color: UIColor.lightGrayColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        cad3.position = bottomCenter
        self.addChild(cad3)
        
        var cad4 =  nodeWithSkill("CAD4", color: UIColor.lightTextColor(), fontSize: 40, fontColor: UIColor.blackColor(),radius: radius)
        cad4.position = bottomRight
        self.addChild(cad4)
        
        
        
    }
    
    let node_Mask : UInt32 = 0b1
    
    func nodeWithSkill(name:String, color:UIColor, fontSize: CGFloat, fontColor: UIColor , radius: CGFloat) -> SKNode{
        let scale = UIScreen.mainScreen().scale
        let screenSize: CGRect = UIScreen.mainScreen().bounds
        let screenWidth = screenSize.width * scale
        let screenHeight = screenSize.height * scale
        
        var node = SKShapeNode()
        var rect = CGRectMake(-radius,-radius,radius*2,radius*2)
        node.name = "CircleNode"
        node.path = UIBezierPath(ovalInRect: rect).CGPath
        node.fillColor = color
        node.strokeColor = UIColor.blackColor()
        node.position = CGPointMake(screenWidth/2, screenHeight / 2)
        node.physicsBody = SKPhysicsBody(circleOfRadius: radius)
        node.physicsBody?.categoryBitMask = node_Mask
        node.physicsBody?.contactTestBitMask = node_Mask
        node.physicsBody?.affectedByGravity = false
        node.physicsBody?.allowsRotation = false
        node.physicsBody?.linearDamping = 1.3
        node.zPosition = 1
        
        var label = SKLabelNode(fontNamed: "ArialMT")
        label.position = CGPointMake(0, 0)
        label.name = "LabelNode"
        label.text = name;
        label.fontSize = fontSize
        label.fontColor = fontColor
        label.userInteractionEnabled = false
        label.horizontalAlignmentMode = SKLabelHorizontalAlignmentMode.Center
        label.verticalAlignmentMode = SKLabelVerticalAlignmentMode.Center
        node.addChild(label)
        
        self.allNodes.append(node)
        return node
    }
    
    var touchesBegan: Bool = false
    override func  touchesBegan(touches: Set<NSObject>, withEvent event: UIEvent) {
        self.touchesBegan = true
        
        for touch in Array(touches) as! [UITouch] {
            var location = touch.locationInNode(self)
            var touchedNode = self.nodeAtPoint(location)
            
            println(touchedNode.name)
        }
    }
    
    var timeInterval :NSTimeInterval = NSTimeInterval()
    var lastTouchTime: NSDate = NSDate()
    
    override func touchesMoved(touches: Set<NSObject>, withEvent event: UIEvent) {
        var now = NSDate();
        self.timeInterval = now.timeIntervalSinceDate(self.lastTouchTime)
        self.lastTouchTime  = NSDate();
        
        
        for touch in Array(touches) as! [UITouch] {
            var location = touch.locationInNode(self)
            var touchedNode = self.nodeAtPoint(location)
            if(touchedNode.name == "LabelNode"){
                touchedNode = touchedNode.parent!
            }
            if(touchedNode.name == "CircleNode"){
                var ptTouch = touch.locationInNode(self)
                var   ptPrevious = touch.previousLocationInNode(self)
                var xMove = ptTouch.x - ptPrevious.x;
                var yMove = ptTouch.y - ptPrevious.y;
                var distance = sqrt ((xMove * xMove) + (yMove * yMove));
                
                touchedNode.position = location;
                touchedNode.physicsBody?.velocity = CGVectorMake(round(xMove) * 30, round(yMove) * 30)
            }
        }
        
        
    }
    
    override func  touchesEnded(touches: Set<NSObject>, withEvent event: UIEvent) {
        self.touchesBegan = false
        startUpdating = true

        //        for touch in Array(touches) as! [UITouch] {
        //            var location = touch.locationInNode(self)
        //            var touchedNode = self.nodeAtPoint(location)
        //            if (touchedNode.name == "LabelNode"){
        //                touchedNode = touchedNode.parent!
        //            }
        //            if (touchedNode.name == "CircleNode"){
        //                if (fabs(touchedNode.physicsBody!.velocity.dx) > 10 || fabs(touchedNode.physicsBody!.velocity.dy) > 10){
        //
        //
        //                    println("Moving")
        //
        //
        //
        //                }
        //
        //            }
        //
        //            
        //        }
        
    }
    
    var resettingNodes: Bool = false
    
    func resetAllNodes(){
        if resettingNodes{
            return
        }else{
            resettingNodes = true
        }
        
        
        let scale = UIScreen.mainScreen().scale
        let screenSize: CGRect = UIScreen.mainScreen().bounds
        let screenWidth = screenSize.width * scale
        let screenHeight = screenSize.height * scale
        
        let bottomLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 1 / 5)
        let bottomCenter = CGPointMake(screenWidth * 2/4, screenHeight * 1 / 5)
        let bottomRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 1 / 5)
        let middleLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 2 / 5)
        let middleCenter = CGPointMake(screenWidth * 2/4, screenHeight * 2 / 5)
        let middleRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 2 / 5)
        let topLeft = CGPointMake(screenWidth * 1/4 - 20, screenHeight * 3 / 5)
        let topCenter = CGPointMake(screenWidth * 2/4, screenHeight * 3 / 5)
        let topRight = CGPointMake(screenWidth * 3/4 + 20, screenHeight * 3 / 5)
        
        var allInPosition = true
        for node in self.allNodes {
            if(!checkIfInPosition(node)){
                allInPosition = false
            }
        }
        
        if(allInPosition){
            resettingNodes = false
            return
        }

        self.allNodes.sort({$0.position.y > $1.position.y })
        var bottomNodes = self.allNodes[6...8]
        var middleNodes = self.allNodes[3...5]
        var topNodes = self.allNodes[0...2]


        
        bottomNodes.sort({$0.position.x < $1.position.x})
        middleNodes.sort({$0.position.x < $1.position.x})
        topNodes.sort({$0.position.x < $1.position.x})
        
        let duration = 1.0
        
        var moveTopLeft = SKAction.moveTo(topLeft, duration: duration)
        var moveTopRight = SKAction.moveTo(topRight, duration: duration)
        var moveTopCenter = SKAction.moveTo(topCenter, duration: duration)
        
        var moveMiddleLeft = SKAction.moveTo(middleLeft, duration: duration)
        var moveMiddleRight = SKAction.moveTo(middleRight, duration: duration)
        var moveMiddleCenter = SKAction.moveTo(middleCenter, duration: duration)
        
        var moveBottomLeft = SKAction.moveTo(bottomLeft, duration: duration)
        var moveBottomRight = SKAction.moveTo(bottomRight, duration: duration)
        var moveBottomCenter = SKAction.moveTo(bottomCenter, duration: duration)
        

        
        
        bottomNodes[0].runAction(moveBottomLeft, completion: {
            self.resettingNodes = false
        })
        bottomNodes[1].runAction(moveBottomCenter)
        bottomNodes[2].runAction(moveBottomRight)
        
        middleNodes[0].runAction(moveMiddleLeft)
        middleNodes[1].runAction(moveMiddleCenter)
        middleNodes[2].runAction(moveMiddleRight)
        
        topNodes[0].runAction(moveTopLeft)
        topNodes[1].runAction(moveTopCenter)
        topNodes[2].runAction(moveTopRight)
    }
    
    var allPositions = [CGPoint]()
    
    func checkIfInPosition(node:SKNode)->Bool{
        var inPosition = false;
        let scale = UIScreen.mainScreen().scale
        let screenSize: CGRect = UIScreen.mainScreen().bounds
        let screenWidth = screenSize.width * scale
        let screenHeight = screenSize.height * scale
        
        for point in allPositions{
            var xDist = fabs(node.position.x - point.x)
            var yDist = fabs(node.position.y - point.y)
            var distance = sqrt((xDist * xDist) + (yDist * yDist))
            if(distance < 5){
                inPosition = true
            }
        }
        
        return inPosition
    }
    
    func movementEnded()->Bool{
        var movement = false
        for node in self.allNodes{
            if(fabs(node.physicsBody!.velocity.dx) > 10 || fabs(node.physicsBody!.velocity.dy) > 10){
                movement = true
            }
        }
        
        
        return movement
    }
    
    var startUpdating:Bool = false
    override func update(currentTime: NSTimeInterval) {
        if (!movementEnded() && startUpdating){
            resetAllNodes()
        }
    }
    
    
    func ExitVC(){
        VC?.dismissViewControllerAnimated(true, completion: nil)
    }
    
}
